source/common/router/router.cc

```diff
@@ -738,6 +778,47 @@ void Filter::sendNoHealthyUpstreamResponse() {
                              StreamInfo::ResponseCodeDetails::get().NoHealthyUpstream);
 }
 
+/*
+ * Helper class to handle a known number of buffer copies.
+ */
+class ManyCopiedBuffer {
+public:
+  ManyCopiedBuffer(Buffer::Instance& original_buffer, int num_copies)
+      : original_buffer_(original_buffer), num_copies_(num_copies) {}
+
+  // Get an unowned reference to the next copy of the buffer. The copy is owned by this class.
+  Buffer::Instance& nextBuffer() {
+    ASSERT(num_copies_ > -1);
+    if (num_copies_ == 0) {
+      --num_copies_;
+      return original_buffer_;
+    } else {
+      copy_ = Buffer::OwnedImpl(original_buffer_);
```

Comment:
we're losing information such as any accounts the original buffer had, we should preserve it (at least for the shadow client use case).

It might make more sense to have the shadow stream itself wired with this.

Reply:
Reasonable, but FWIW the existing shadowing approach doesn't do any accounting either. Would you be ok if we postponed that to a future PR?